### PR TITLE
[Inductor] add expm1 lowering

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89956
* __->__ #89961

Improves perf of inductor no-cudagraphs on nvidia-deeprecommender from 0.88 -> .96. I am looking into disabling implicit fallbacks for benchmark models in another pr. 

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/decomposition.py [hunk 1]
```diff
@@ -411,6 +411,9 @@ def nan_to_num(x, nan=0.0, posinf=None, neginf=None):
 def all(input):
     return torch.logical_not(torch.any(torch.logical_not(input)))
 
+@register_decomposition([aten.expm1.default])
+def expm1(input):
```
Comment (ngimel) [inline torch/_inductor/decomposition.py:415]: We need to do a lowering with a libdevice/std function (similar to erf), this is not enough accuracy. While you are at it, can you also delete log1p decomp from inductor and do the same lowering thing? With the tests for values close to one to check accuracy.

Comment (eellison) [reply torch/_inductor/decomposition.py:415]: I looked, and `expm1` isn't listed in the [ptx instruction set](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html). What am I missing ?

Comment (eellison) [reply torch/_inductor/decomposition.py:415]: nvm it's here https://docs.nvidia.com/cuda/libdevice-users-guide/__nv_expm1.html

Comment (ngimel) [reply torch/_inductor/decomposition.py:415]: It's not in ptx, it's in triton's wrapper for libdevice that we call like this https://github.com/pytorch/pytorch/blob/master/torch/_inductor/codegen/triton.py#L211 for a lot of functions that are not in ptx set.

**Review by ngimel — APPROVED** [review]
Awesome, thanks!

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -2693,6 +2693,34 @@ def fn(x):
...
+            return torch.expm1(x), torch.expm1(x) * 2
+
+        for dtype in (torch.float16, torch.float, torch.double, torch.int, torch.int64):
+            self.common(
+                fn,
+                (torch.randn([64]).to(dtype=dtype),),
+            )
+            self.common(
+                fn,
+                (torch.arange(-0.00001, 0.00001, 0.001).to(dtype=dtype),),
```
Comment (ngimel) [inline test/inductor/test_torchinductor.py:2707]: the step here is too large, you end up with a single value, -1e-5

Comment (eellison) [reply test/inductor/test_torchinductor.py:2707]: oops, good catch

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -400,6 +400,14 @@ def sqrt(x):
     def rsqrt(x):
         return f"1 / std::sqrt({x})"
 
+    @staticmethod
+    def log1p(x):
```
Comment (desertfire) [inline torch/_inductor/codegen/cpp.py:404]: @jgong5, can these be vectorized?
